### DISCUSSION TITLE
Interesting Collision issues

### DISCUSSION THREAD

**Discussion by xLink** [Annoraaq/grid-engine#184] [General]
hey, so im using the engine, to clone a pokemon game, so far so good, got lots of basic things working, including a custom follow thing for the pokemon. given the img:
![image](https://user-images.githubusercontent.com/217433/130439684-a24baa83-1b5e-4615-bfe8-ecffbbac99c2.png)

The player is supposed to move left to right across the bridge, ive added a `ge_collide_down` and `ge_collide_up` to the bridge tiles, the edge of the rocks are full `ge_collide` (which is stopping the player in the img), but the player should also be able to move underneath the bridge in the img too...how do i handle the collision here?

feels like i need collision based on what layer or tile the character is stood on but my head is hurting trying to figure this out XD

**Comment by Annoraaq** [comment]
That is indeed an interesting problem, not only for collisions. Another problems is the display of the tile layers because the bridge should suddenly be rendered on top of the player.

First of all, I fear with the current functionality of the plugin there is no built in way to solve this.

To solve the display layer and collision problems I suggest the following features for GridEngine:

We introduce GridEngine layers. Each GridEngine layer gets the two layers of the tilemap it should appear inbetween. So if you want your GridEngine chars of layer x to appear between tilemap layer 2 and 3, you can specify it while configuring GridEngine.

GridEngine needs to use several own layers for correct character positioning. So it will adapt the depth settings of the tilemap and character sprites accordingly.

So in your case you would create two GridEngine layers. One on top of the bridge and one below. In your code you need to make sure that your character switches layers when walking down/up the ramp/stairs.

Regarding the collision problems:

You are not the only one demanding a more granular control over collisions. There were collision layers in the room: https://github.com/Annoraaq/grid-engine/discussions/181#discussioncomment-1214202

However, I suggest to use collision groups instead of layers. Layers are only needed if the order is relevant which is probably not the case for collisions. Each character can be part of zero or more collision groups. In the tilemap each tile can be part of zero or more collision groups.

In your case you could then have a collision group for the ground floor and one for being on top of the bridge. When a character passes the ramp/stairs you need to make it switch the GridEngine layer AND the collision group.

What do you think of that approach?

**Reply by xLink** [reply]
> So if you want your GridEngine chars of layer x to appear between tilemap layer 2 and 3, you can specify it while configuring GridEngine.

will this allow for characters to be in multiple layers at a time? eg if i put one on the bridge and one underneath, they would be fine?

> You are not the only one demanding a more granular control over collisions. There were collision layers in the room: #181 

very much not demanding, but it is the only way to make more refined games with the engine, other than the docs, do you know of any decent sized projects that are using this?

> When a character passes the ramp/stairs you need to make it switch the GridEngine layer AND the collision group.

does it make sense for these to be tile properties that could automatically make the character switch?

**Reply by Annoraaq** [reply]
> will this allow for characters to be in multiple layers at a time? eg if i put one on the bridge and one underneath, they would be fine?

Yes that would be fine. Only the same character can be on only one layer at a time. This constraint is not valid for collision groups, though. There it can make sense to have the same character be part of several collision groups.

> very much not demanding, but it is the only way to make more refined games with the engine, other than the docs, do you know of any decent sized projects that are using this?

Maybe not the best wording from my side. I just meant some feature like this is helpful to your problem and you are not the only one who would profit from it.
Unfortunately I am not aware of most of the projects this is being used. So if you find out any decent sized one, please let me know :D

> does it make sense for these to be tile properties that could automatically make the character switch?

I think that would be a good idea. In the first iteration I would probably not include it to the plugin so you would have to create it yourself (which should not be a big problem). But in a later version it might make sense to have some tile property that increases/decreases the GridEngine layer when stepped on.

**Reply by xLink** [reply]
i dont mind being a guinea pig for this functionality, and hopefully my project should be of decent size if i can get it done XD
not trying to rush, just wondering i fwe have a timescale on this being implemented? i have a ton of other things i can do in the mean time

**Reply by Annoraaq** [reply]
It has the highest priority at the moment. If no urgent bug shows up it should be doable in few weeks.

**Reply by xLink** [reply]
> Unfortunately I am not aware of most of the projects this is being used. So if you find out any decent sized one, please let me know :D

just a quick Q, are you on discord, if so add me `xLink#2097` would be great to get some insight to how im setting some of this up XD

**Reply by Annoraaq** [reply]
just sent you an invite
